feat(mistral): add FIM support for codestral models

markijbema opened the pull request:
- Add supportsFim(), completeFim(), and streamFim() methods to MistralHandler
- FIM support is only enabled for codestral models (codestral-latest, codestral-2405, etc.)

Commit:
feat(mistral): add FIM support for codestral models

- Add supportsFim(), completeFim(), and streamFim() methods to MistralHandler
- FIM support is only enabled for codestral models (codestral-latest, codestral-2405, etc.)
- Uses Mistral's v1/fim/completions endpoint with proper base URL routing
- Supports both streaming and non-streaming FIM completions
- Add comprehensive test coverage with 10 test cases
- Update continuedev Mistral LLM to accurately check model support
- Implementation follows the same pattern as KiloCode OpenRouter handler

Closes: FIM functionality now works for Mistral's codestral models

markijbema commented:
this works, but you have to change the base url in the provider, and ask specific access with a different key

<img width="2528" height="1072" alt="CleanShot 2025-11-21 at 11 44 29@2x" src="https://github.com/user-attachments/assets/131e2d37-b0c5-421b-bc3d-b5bb609dfdb3" />

Commit:
feat(ghost): use duck typing for FIM support in GhostModel

- Add FimCapableHandler interface for duck typing
- Add isFimCapable() type guard function
- Update supportsFim() and generateFimResponse() to use duck typing
- Any handler implementing supportsFim() and streamFim() now works with FIM
- Enables MistralHandler FIM support without explicit class checks

Commit:
fix: add optional taskId and onUsage params to MistralHandler.streamFim

Address review comment about signature mismatch with IFimProvider interface.
The streamFim method now accepts optional taskId and onUsage parameters,
and calls onUsage when usage data is available in the stream response.

markijbema commented:
Addressed the review comment about the signature mismatch with `IFimProvider` interface.

The `MistralHandler.streamFim` method now:
1. Accepts optional `taskId` and `onUsage` parameters to match the interface signature
2. Calls `onUsage` when usage data is available in the stream response (Mistral FIM API returns `prompt_tokens`, `completion_tokens`, and `total_tokens` in the final chunk)

This enables usage tracking for Mistral FIM completions when the callback is provided.

Commit:
add missing markers

chrarnoldus commented on src/api/providers/mistral.ts:
```diff
@@ -209,4 +212,81 @@ export class MistralHandler extends BaseProvider implements SingleCompletionHand
+		return modelId.startsWith("codestral-")
+	}
+
+	async completeFim(prefix: string, suffix: string, _taskId?: string): Promise<string> {
```
Is this used outside of tests?

Commit:
Remove unused completeFim method from IFimProvider interface

- Remove completeFim from IFimProvider interface
- Remove completeFim implementation from MistralHandler
- Remove completeFim implementation from KilocodeOpenrouterHandler
- Remove completeFim tests from both test files

The method was only used in tests and not in production code.
streamFim is the only FIM method needed.

Commit:
remove markers

add new file marker to hopefully prevent further markers

markijbema commented on src/api/providers/mistral.ts:
```diff
@@ -209,4 +212,81 @@ export class MistralHandler extends BaseProvider implements SingleCompletionHand
+		// Get the base URL for the model
+		const baseUrl = model.startsWith("codestral-")
+			? this.options.mistralCodestralUrl || "https://codestral.mistral.ai"
+			: "https://api.mistral.ai"
```
no

Commit:
fix types and added comment

Commit:
add changeset
